### PR TITLE
New CVaR atom

### PR DESCRIPTION
**Pull request by davidppineiro**
## Description
This PR creates the `cvar` atom (conditonal value at risk), which is implemented as a wrapper around the existing `dotsort` atom.

## Type of change
- [X] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [X] Add our license to new files.
- [X] Check that your code adheres to our coding style.
- [X] Write unittests.
- [X] Run the unittests and check that they’re passing.
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

**Comment by PTNobel** [comment]
Hey David! I'd encourage that you install and setup the pre-commit hook to make sure your files are properly formatted.

Overall seems like a fine change, but we'll need tests and docs to go with it too.

Tests can be whatever we have for dotsort, just a couple quick tests, and docs can just be an additional row in the atoms table.

**Comment by Transurgeon** [comment]
Hello @davidppineiro thank you for your contribution. 
Just reiterating on Parth's suggestion and giving you more pointers.
- code style instructions can be found [here](https://github.com/cvxpy/cvxpy/blob/master/CONTRIBUTING.md#code-style).
- tests should be added in ``test_atoms.py`` below the ``TestDotSort`` class. You could try solving the wikipedia definition of cvar and do an assertion that the cvar atom gives a similar answer (up to some very small precision). 
- docs will mostly involve a better description in the docstrings of the function. Put some math, and explain the type of the inputs (x must be an vector expression, alpha must be a float between 0 and 1, representing a percentage), etc. Then to add it to the website, you would need to add one line below ``cp.cummax`` [here](https://github.com/cvxpy/cvxpy/blob/master/doc/source/api_reference/cvxpy.atoms.other_atoms.rst#cummax) so that sphinx can generate the docs for the new cvar atom.

**Comment by PTNobel** [comment]
> docs will mostly involve a better description in the docstrings of the function. Put some math, and explain the type of the inputs (x must be an vector expression, alpha must be a float between 0 and 1, representing a percentage), etc. Then to add it to the website, you would need to add one line below cp.cummax [here](https://github.com/cvxpy/cvxpy/blob/master/doc/source/api_reference/cvxpy.atoms.other_atoms.rst#cummax) so that sphinx can generate the docs for the new cvar atom.

It also needs to be added to the Atoms table! David, you can find me on Tuesday and I can show you where that goes.

**Review by SteveDiamond — APPROVED** [review]
Nice!

File: doc/source/tutorial/functions/index.rst [hunk 1]
```diff
@@ -81,6 +81,15 @@ and returns a scalar.
      - Curvature |_|
      - Monotonicity
 
+  * - :ref:`cvar(x, beta) <cvar>`
+      - average of the :math:`(1-\beta)` fraction of worst outcomes in :math:`x`
```
Comment (PTNobel) [inline doc/source/tutorial/functions/index.rst:85]: @Transurgeon Is this too long for the new format?

Comment (Transurgeon) [reply doc/source/tutorial/functions/index.rst:85]: yea most certainly will, good catch.
I think we might need to separate it into two lines.
Let met try to generate the docs and see how it looks.

Comment (Transurgeon) [reply doc/source/tutorial/functions/index.rst:85]: ok, there was actually a formatting error and the table wasn't showing at all.. good that we checked before merging.
@PTNobel or @davidppineiro , can you change the atoms table change from line 84-92 to the following (make sure the star and dashes are aligned with everything else)?:
```py
   * - :ref:`cvar(x, beta) <cvar>`
     - average of the :math:`(1-\beta)`
      
       fraction of worst
       
       outcomes in :math:`x`
     - :math:`x \in \mathbf{R}^m`
      
       :math:`\beta \in (0,1)`
     - depends on :math:`x`
     - |convex| convex
     - |incr| incr.
```

Comment (davidppineiro) [reply doc/source/tutorial/functions/index.rst]: Done! Does it show up now?

File: doc/source/tutorial/functions/index.rst [hunk 2]
```diff
@@ -81,6 +81,19 @@ and returns a scalar.
      - Curvature |_|
      - Monotonicity
 
+   * - :ref:`cvar(x, beta) <cvar>`
+     - average of the :math:`(1-\beta)`
+      
+       fraction of worst
```
Comment (PTNobel) [inline doc/source/tutorial/functions/index.rst:87]: Last change, what does worst mean? Can we either say smallest or largest?

**Review by PTNobel — CHANGES_REQUESTED** [review]

**Review by Transurgeon — APPROVED** [review]
LGTM on condition of addressing Parth's suggestion.
Thanks for doing this @davidppineiro !

File: doc/source/tutorial/functions/index.rst (hunk 2 above)
Comment (davidppineiro) [reply doc/source/tutorial/functions/index.rst:87]: If we go with largest/smallest, we have to specify if the vector x represents losses or profits. 
Which version do you prefer?
1. average of the :math:`(1-\beta)` fraction of worst values in :math:`x`
2. average of the :math:`(1-\beta)` fraction of largest losses in :math:`x`
Let me know if you have a better alternative.

Comment (SteveDiamond) [reply doc/source/tutorial/functions/index.rst:87]: +1 for option 3

File: doc/source/tutorial/functions/index.rst [hunk 3]
```diff
@@ -83,17 +83,19 @@ and returns a scalar.
...
      - average of the :math:`(1-\beta)`
-      
-       fraction of worst
        
-       outcomes in :math:`x`
+       fraction of largest values
+       
+       in :math:`x`, where :math:`x`
+       
+       is a vector of losses
```
Comment (SteveDiamond) [inline doc/source/tutorial/functions/index.rst:91]: do we need to say what `x` is, given we are saying cvar is averaging the largest values?

Comment (SteveDiamond) [reply doc/source/tutorial/functions/index.rst:91]: in other words I don't think we need the part about `where x is a vector of losses`

Comment (davidppineiro) [reply doc/source/tutorial/functions/index.rst:91]: Yeah, I think you are right. It's redundant. Will take it out.

File: doc/source/tutorial/functions/index.rst (hunk 2 above)
Comment (PTNobel) [reply doc/source/tutorial/functions/index.rst]: Definitely 3

**Comment by davidppineiro** [comment]
@PTNobel @SteveDiamond I have made some changes to the docstring in the CVaR atom and the table, and included some checks. Let me know if you have more suggestions.

File: cvxpy/atoms/cvar.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    Returns
+    -------
+    Expression
+        The CVaR of :math:`x` at probability level :math:`\beta`.
+    """
+    if not 0 <= beta < 1:
+        raise ValueError(f"The probability level beta must be in the range [0, 1), got {beta}")
+
+    if len(x.shape) > 1 and x.shape[1] != 1:
```
Comment (SteveDiamond) [inline cvxpy/atoms/cvar.py:52]: I don't think this is the condition we want. It should be `len(x.shape) == 1`.

Comment (davidppineiro) [reply cvxpy/atoms/cvar.py:52]: But that will fail for x = cp.Variable((n,1)). Is this what we want?

Comment (SteveDiamond) [reply cvxpy/atoms/cvar.py:52]: Yes, that should fail. It's not consistent with numpy style to allow `x = cp.Variable((n,1))` but not allow `x = cp.Variable((1,n))`. If we allow 2D inputs we need to start thinking about having an axis argument, which would be much more complicated (though maybe doable?)

Comment (davidppineiro) [reply cvxpy/atoms/cvar.py]: Makes sense. Just changed the assert to raise an error if it's not a 1D array.

**Comment by davidppineiro** [comment]
@SteveDiamond @PTNobel Also added another test for the CVaR atom: solving a small LP with a CVaR constraint.

**Review by SteveDiamond — APPROVED** [review]
Nice job!

**Review by PTNobel — APPROVED** [review]

**Pull request merged**